lib.systems: deprecate `==` comparison between elaborated platforms

MattSturgeon opened the pull request:
As suggested by @wolfgangwalther https://github.com/NixOS/nixpkgs/pull/380005#discussion_r1947114008, I've scaled back this PR to focus on the treewide deprecation of `==` comparison between elaborated platform attrsets. The other changes are replaced by #380342.

This is in contrast to the previous effort to fully support `==` with elaborated platform attrsets: #238331. Also related: #278001.

A "backwards compatibility" item has been added to the release notes.

Added `equals` and `notEquals` functions to elaborated systems. This makes replacing `==` and `!=` usage simpler, and also provides a convenient alternative to using `lib.systems.equals` directly.

Added an extra attr (`warnAboutEquality`) to elaborated systems that exists only to produce a warning when evaluated. This is intended to detect `==` comparison being used with platform attrsets, suggested by @roberth https://github.com/NixOS/nixpkgs/pull/380005#discussion_r1946245110.

This approach isn't perfect, because it is only evaluated if the `==` comparison has not already short-circuited. E.g. if the attr-count is different or some other difference has already been found before getting to the `warnAboutEquality` attr. These limitations were raised by @wolfgangwalther https://github.com/NixOS/nixpkgs/pull/380005#discussion_r1947132275 and were also discussed in [this note](https://github.com/NixOS/nixpkgs/pull/238331#issuecomment-1600066095).

The treewide change has been done using some regex search/replace together with vim's quickfix list and some manual checking. It covers ~600 changes, so I haven't been able to check each individual replacement in detail. As mentioned by @roberth, many of the replacements could probably use `canExec` instead of `equals`, but that change is not in scope for this PR.


<details><summary>Old summary</summary>
<p>

Intended to enable composition of package sets by preserving the original un-elaborated arguments passed to `lib.systems.elaborate`.

This PR does not actually make package sets composable, but it is intended that the implementation proposed here will enable @wolfgangwalther to revisit their earlier work in #303849 and #376988 without needing to make changes to the public API or the `nixpkgs` nixos module.

Adds:
- `lib.systems.isElaborated`
- A `_type` tag for elaborated systems
- `input` attr on elaborated systems
- `lib.systems.equals` ignores the new `input` attr

Additionally:
- added `equals` and `notEquals` "method" attr-functions for convenience, and to allow comparison without needing access to `lib`
- treewide migration from `==` -> `<platform>.equals`
- elaborated platforms now warn when fully evaluated, in order to warn against using `==`


</p>
</details> 

cc @wolfgangwalther @roberth

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [x] `nix-build ./lib/tests/release.nix --abort-on-warn --show-trace`
  - [ ] [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - [ ] and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - [ ] or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - [ ] made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth commented on lib/systems/default.nix:
```diff
@@ -36,14 +36,18 @@ let
     compare the value with a reconstruction of itself, e.g. with `f == a: f a`,
     or perhaps calling `elaborate` twice, and one will see reflexivity fail as described.
 
+    Additionally, elaborated systems contain their original "input" args. These
+    should be ignored when checking equality because multiple inputs could produce
+    the same elaborated systems.
+
```
We have previously tried to make them comparable via `==`
- https://github.com/NixOS/nixpkgs/pull/238331

Currently such comparisons only work for the exact ones in a single `pkgs` instance, because they are deduplicated, making them identical instead of just equal (which they wouldn't be thanks to a pointer comparison quirk).

This change would put the final nail in the coffin for platform equality. Whereas with a canonicalization function like `elaborate` this is achievable, if you preserve the original input, you'll have an equality that works most of the time, but reports `false` when equal platforms are constructed with equivalent syntaxes.

A lot of code currently relies on this equality, which works by virtue of the deduplication, which we can keep doing, but I do believe that this may be an anti-pattern. Usually `canExecute` would be more accurate.

If we go this route, it would be a good idea to start deprecating equality, which can more or less be done by adding an attribute that prints a warning when its value is evaluated.

```nix
# top of file
let
  # top-level binding => only print once
  warnSystemEquals = lib.warn "It seems that you're comparing platform attrsets using plain == equality, which is not supported, and will result in an error soon. Use `--abort-on-warn` to find the location of this warning. If equality (a == b) was used, try (a.canExecute b) instead. If equality was not used, you may avoid this warning by filtering out the attribute `warnAboutEquality`." null;
in
# rest of systems.nix

  elaborate = 
    # ...
    <result> // {
      warnAboutEquality = warnSystemEquals;
    };
```

MattSturgeon replied:
I was unsure whether it'd be preferred to keep such de-duplication and preserve referential equality (`==`), but of course this would be at the cost of loosing `input` for the de-duplicated platform.

It sounds like it'd be preferred to fully preserve all `input`s, at the cost of deprecating `==` comparison? So I'll add the warning you proposed, thanks for the suggestion.

MattSturgeon replied:
Adding the warning required a treewide change. Should I split that out into a dedicated PR?

roberth replied:
It kind requires that we use the **changes from** dropdown in the GitHub PR files UI, which is not great, but better than breaking CI.

roberth commented on lib/systems/default.nix:
```diff
@@ -69,6 +78,9 @@ let
     allArgs = systemToAttrs systemOrArgs;
 
+    # Preserve the original input arguments
+    input = if isElaborated allArgs then allArgs.input else allArgs;
```
Good to at least remove the entropy of string vs `.system` string. :+1:

roberth commented on lib/systems/default.nix:
```diff
@@ -421,6 +433,9 @@ let
             any (t: hasInfix t final.rust.rustcTarget) ["-none" "nvptx" "switch" "-uefi"];
         };
+      } // {
+         _type = "elaborated-system";
```
I call these "platform objects", but that might just be me. They're definitely not system strings though. Having a `system` in a system object would be confusing to talk about, especially in actually vocal speak, in person.
```suggestion
         _type = "platform";
```



I don't think non-elaborated ones will ever be type-tagged, but if you insist, you may still consider

```suggestion
         _type = "elaborated-platform";
```

MattSturgeon replied:
I was kinda unsure how to name it myself. It seemed "system" and "elaborated system" was used internally. But I agree, "platform" is clearer IMO.

roberth reviewed (commented):
My thoughts. I'm not the most knowledgeable person on the subject of these platform values, but I'd like to support you and @wolfgangwalther if possible.

MattSturgeon committed:
lib.systems.elaborate: ensure platforms warn when compared using `==`

Added `warnAboutPlatformEquals` to elaborated platforms.

This new attr prints a warning when evaluated. This is intended to
detect `==` comparison being used with platform attrsets.

This isn't perfect, because it is only evaluated if the `==` comparison
has not already short-circuited. E.g. if the attr-count is different or
some other difference has already been found before getting to the
`warnAboutPlatformEquals` attr.

Added a "backwards compatibility" item to the release notes.

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

MattSturgeon committed:
treewide: use `<platform>.equals` not `==`

In preparation for deprecating `==` usage on platform attrs.

MattSturgeon committed:
lib.systems.elaborate: add `equals` and `notEquals` attrs

Allows replacing `lib.systems.equals foo bar` with `foo.equals bar`.

This is useful when `lib` isn't readily available, and is often less
verbose.

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 380005`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

roberth commented on lib/systems/default.nix:
```diff
@@ -20,6 +21,15 @@ let
+    "It seems that you're comparing platform attrsets using plain == equality,"
+    "which is not supported, and will result in an error soon."
+    "Use `--abort-on-warn` to find the location of this warning."
+    "If equality (a == b) was used, try (a.canExecute b) instead."
```
```suggestion
    "If equality (e.g. host == build) was used, try (build.canExecute host) instead."
```

roberth commented on lib/systems/default.nix:
```diff
@@ -20,6 +21,15 @@ let
+  warnSystemEquals = lib.warn (lib.concatStringsSep " " [
+    "It seems that you're comparing platform attrsets using plain == equality,"
+    "which is not supported, and will result in an error soon."
+    "Use `--abort-on-warn` to find the location of this warning."
```
```suggestion
    "Use `--abort-on-warn --show-trace` to find the location of this warning."
```

roberth commented on nixos/modules/config/fonts/fontconfig.nix:
```diff
@@ -58,7 +58,7 @@ let
         <!-- Font directories -->
         ${lib.concatStringsSep "\n" (map (font: "<dir>${font}</dir>") config.fonts.packages)}
-        ${lib.optionalString (pkgs.stdenv.hostPlatform == pkgs.stdenv.buildPlatform) ''
+        ${lib.optionalString (pkgs.stdenv.hostPlatform.equals pkgs.stdenv.buildPlatform) ''
```
Most of these could be changed to `buildPlatform.canExecute hostPlatform`, but let's keep the treewide change minimal.

MattSturgeon replied:
I don't feel qualified to know when `canExecute` is more suitable than `equals`, and I can't think or a reasonable way to scale that judgement call across ~600 replacements.

But I'm happy to try and improve the treewide if you have any ideas for automating the replacement.

I guess I could change all matches to use `canExecute`, but IDK how easy it'd be to verify that hasn't broken anything.

Therefore I'm minded to **agree**: keep the treewide as simple as possible :grin: 

<details><summary>Treewide methodology so far</summary>
<p>


For the changes so far, I used a regex to find each `[^ ]Platform ==` and then used vim's quickfix list to go through and confirm each one:

```
:cdo s/\(\<[^( ]\+Platform\) ==/\1.equals/cg
```

There were ~200matches for `==` and ~400 for `!=`.


</p>
</details>

alyssais replied:
I would also prefer using equals for now.

roberth commented on lib/systems/default.nix:
```diff
@@ -20,6 +21,15 @@ let
   inherit (lib.strings) toJSON;
 
+  # define this binding in the top-level let-block so that it only prints once
+  warnSystemEquals = lib.warn (lib.concatStringsSep " " [
```
I guess for consistency:

```suggestion
  warnPlatformEquals = lib.warn (lib.concatStringsSep " " [
```

roberth reviewed (commented):
Have a few suggestions, but other than that, lgtm.
Would appreciate someone else to also review this.

MattSturgeon commented on lib/systems/default.nix:
```diff
@@ -421,6 +449,11 @@ let
         };
+      } // {
+        _type = "platform";
+        inherit input warnPlatformEquals;
```
That sounds like a good suggestion to me. I can't think of any reason why an outside user would need access to the actual inputs if they had an `override` function.

Although, if we used `mkOverridable`, the inputs would be available via a `__functionArgs` attr on the `override` functor. So still accessible if needed.

MattSturgeon commented on lib/systems/default.nix:
```diff
@@ -55,6 +75,11 @@ let
   */
   flakeExposed = import ./flake-systems.nix { };
 
+  /**
+    Whether `x` is an elaborated system/platform.
+  */
+  isElaborated = x: x._type or null == "platform";
+
```
Sure. I think I still had in my head the nixpkgs module wanting to check whether inputs were already elaborated. But you're right, those kinda public-facing changes shouldn't be necessary with this approach.

MattSturgeon commented on nixos/doc/manual/release-notes/rl-2505.section.md:
```diff
@@ -176,6 +176,10 @@
   instead of the python tester launcher. You can still refer to the python
   launcher via `python3Packages.toPythonApplication python3Packages.playwright`
 
+- Using `==` or `!=` to compare elaborated platform attrsets, such as `stdenv.hostPlatform` and `stdenv.buildPlatform`, is no longer supported and may produce a warning. In a future version, this warning will be replaced with an error. Instead, you should use the `canExecute`, `equals`, or `notEquals` functions included with platform attrsets or use `lib.systems.equals`.
+
+  - E.g. replace `hostPlatform == buildPlatform` with `buildPlatform.canExecute hostPlatform`.
```
Added a release note to the "Backward Incompatibilities" section.

Alternative wording suggestions welcome :upside_down_face:

MattSturgeon replied:
Hm, should this go in [`doc/release-notes`](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2505.section.md) or [`nixos/doc/manual/release-notes`](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md)? Maybe both?

MattSturgeon replied:
Thanks. Moved :+1:

MattSturgeon committed:
lib/tests/systems.nix: simplify `test_equals_unequal` using removeAttrs

A `removeAttrs` call was added by 4eca30db7da1aca7b7e91793d54a72739b8a407c
We may as well use it to simplify the test's implementation.

MattSturgeon commented on lib/tests/systems.nix:
```diff
@@ -102,17 +102,22 @@ lib.runTests (
-}) (lib.systems.elaborate "x86_64-linux" /* arbitrary choice, just to get all the elaborated attrNames */)
+  })
+  # x86_64-linux is an arbitrary choice, just to get all the elaborated attrNames
+  (builtins.removeAttrs (lib.systems.elaborate "x86_64-linux") [
+    # Avoid the warning
+    "warnAboutPlatformEquals"
+    # Changes in these attrs are not detectable because they're function.
+    # The functions should be derived from the data, so this is not a problem.
+    "canExecute"
+    "emulator"
+    "emulatorAvailable"
+    "equals"
+    "isCompatible"
+    "notEquals"
+    "staticEmulatorAvailable"
+  ])
```
This PR was adding a `removeAttrs` here anyway, so I've added an extra commit that refactors this test to take advantage of it.

This is a little out of scope, so I can drop the commit if preferred.

MattSturgeon committed:
nixos/nixpkgs: deprecate `==` support for `hostPlatform` & `buildPlatform`

MattSturgeon commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -220,15 +220,7 @@ in
       };
       # Make sure that the final value has all fields for sake of other modules
       # referring to this.
-      apply =
-        inputBuildPlatform:
-        let
-          elaborated = lib.systems.elaborate inputBuildPlatform;
-        in
-        if lib.systems.equals elaborated cfg.hostPlatform then
-          cfg.hostPlatform # make identical, so that `==` equality works; see https://github.com/NixOS/nixpkgs/issues/278001
-        else
-          elaborated;
```
Because out-of-tree users may be doing `config.nixpkgs.hostPlatform == config.nixpkgs.buildPlatform` and this change will break that functionality?

I guess that makes sense.

_My_ thought process was that the current coercion may be incompatible with what your trying to achieve in #380342. But it's probably not a _big_ deal to lose one platform's potentially unique `override` if those platforms are effectively equivalent anyway.

I'll drop the change and add a TODO comment.

MattSturgeon commented:
Rebased to resolve conflicts

nyabinary commented:
Has merge conflicts that need to be resolved because of the treewide format

MattSturgeon commented:
I've just pushed a WIP rebase that resolves merge conflicts. I still need to go through the tree and fix any new instances of `==` or `!=`. Looks like one such usage was [picked up by CI](https://github.com/NixOS/nixpkgs/actions/runs/14266699531/job/39990229861?pr=380005), which is promising.

MattSturgeon committed:
treewide: fix platform-string comparison

`(elaborated platform) == "string"` will always be false. This was
probably intended as `(elaborated platform).system == "string"`.

That said, this should usually be replaced with checking the relevant
boolean attributes, such as `isLinux` or `isDarwin`.

MattSturgeon commented on pkgs/development/python-modules/sentry-sdk/1.nix:
```diff
@@ -116,7 +116,7 @@ buildPythonPackage rec {
       "tests/integrations/"
     ]
-    ++ lib.optionals (stdenv.buildPlatform != "x86_64-linux") [
+    ++ lib.optionals (stdenv.buildPlatform.system != "x86_64-linux") [
       # test crashes on aarch64
       "tests/test_transport.py"
```
Based on the comment, I'm guessing this should be `!buildPlatform.isAarch64`

MattSturgeon replied:
cc authors: @gador and @SuperSandro2000

gador replied:
To be fair, I only tested it on `aarch64-linux` and `x86_64-linux`. I don't know wheter it would fail for other systems (e.g. `x86_64-darwin`). OTOH we can just try to go with `buildPlatform.isAarch64` and see how it works

MattSturgeon replied:
I've changed it to use `buildPlatform.isAarch64` in my last rebase.

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 380005`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>cura</li>
    <li>netbox_3_7</li>
  </ul>
</details>
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 49 packages failed to build:</summary>
  <ul>
    <li>litellm</li>
    <li>litellm.dist</li>
    <li>netbox (netbox_4_2)</li>
    <li>netbox_4_1</li>
    <li>peering-manager</li>
    <li>python312Packages.django-rq</li>
    <li>python312Packages.django-rq.dist</li>
    <li>python312Packages.netbox-bgp</li>
    <li>python312Packages.netbox-bgp.dist</li>
    <li>python312Packages.netbox-contract</li>
    <li>python312Packages.netbox-contract.dist</li>
    <li>python312Packages.netbox-documents</li>
    <li>python312Packages.netbox-documents.dist</li>
    <li>python312Packages.netbox-floorplan-plugin</li>
    <li>python312Packages.netbox-floorplan-plugin.dist</li>
    <li>python312Packages.netbox-interface-synchronization</li>
    <li>python312Packages.netbox-interface-synchronization.dist</li>
    <li>python312Packages.netbox-napalm-plugin</li>
    <li>python312Packages.netbox-napalm-plugin.dist</li>
    <li>python312Packages.netbox-plugin-prometheus-sd</li>
    <li>python312Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python312Packages.netbox-qrcode</li>
    <li>python312Packages.netbox-qrcode.dist</li>
    <li>python312Packages.netbox-reorder-rack</li>
    <li>python312Packages.netbox-reorder-rack.dist</li>
    <li>python312Packages.netbox-routing</li>
    <li>python312Packages.netbox-routing.dist</li>
    <li>python312Packages.netbox-topology-views</li>
    <li>python312Packages.netbox-topology-views.dist</li>
    <li>python312Packages.rq</li>
    <li>python312Packages.rq.dist</li>
    <li>python312Packages.wagtail-localize</li>
    <li>python312Packages.wagtail-localize.dist</li>
    <li>python313Packages.netbox-bgp</li>
    <li>python313Packages.netbox-bgp.dist</li>
    <li>python313Packages.netbox-documents</li>
    <li>python313Packages.netbox-documents.dist</li>
    <li>python313Packages.netbox-interface-synchronization</li>
    <li>python313Packages.netbox-interface-synchronization.dist</li>
    <li>python313Packages.netbox-plugin-prometheus-sd</li>
    <li>python313Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python313Packages.netbox-qrcode</li>
    <li>python313Packages.netbox-qrcode.dist</li>
    <li>python313Packages.netbox-reorder-rack</li>
    <li>python313Packages.netbox-reorder-rack.dist</li>
    <li>python313Packages.netbox-routing</li>
    <li>python313Packages.netbox-routing.dist</li>
    <li>recon-ng</li>
    <li>recon-ng.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 46 packages built:</summary>
  <ul>
    <li>alerta-server</li>
    <li>alerta-server.dist</li>
    <li>frigate</li>
    <li>gns3-gui</li>
    <li>gns3-gui.dist</li>
    <li>gns3-server</li>
    <li>gns3-server.dist</li>
    <li>kdePackages.drkonqi</li>
    <li>kdePackages.drkonqi.debug</li>
    <li>kdePackages.drkonqi.dev</li>
    <li>kdePackages.drkonqi.devtools</li>
    <li>ledfx</li>
    <li>ledfx.dist</li>
    <li>matrix-synapse</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth.dist</li>
    <li>matrix-synapse-unwrapped</li>
    <li>matrix-synapse-unwrapped.dist</li>
    <li>moodle-dl</li>
    <li>moodle-dl.dist</li>
    <li>mozphab</li>
    <li>mozphab.dist</li>
    <li>nixpkgs-manual</li>
    <li>notus-scanner (python312Packages.notus-scanner)</li>
    <li>notus-scanner.dist (python312Packages.notus-scanner.dist)</li>
    <li>octoprint</li>
    <li>octoprint.dist</li>
    <li>ospd-openvas</li>
    <li>ospd-openvas.dist</li>
    <li>protonvpn-gui</li>
    <li>protonvpn-gui.dist</li>
    <li>python312Packages.proton-vpn-api-core</li>
    <li>python312Packages.proton-vpn-api-core.dist</li>
    <li>python312Packages.proton-vpn-network-manager</li>
    <li>python312Packages.proton-vpn-network-manager.dist</li>
    <li>python312Packages.sentry-sdk (python312Packages.sentry-sdk_1)</li>
    <li>python312Packages.sentry-sdk.dist (python312Packages.sentry-sdk_1.dist)</li>
    <li>skytemple</li>
    <li>skytemple.dist</li>
    <li>tribler</li>
  </ul>
</details>

wolfgangwalther reviewed (approved):
Still LGTM.

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 380005`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>cura</li>
    <li>netbox_3_7</li>
  </ul>
</details>
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>python312Packages.wagtail-localize</li>
    <li>python312Packages.wagtail-localize.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 93 packages built:</summary>
  <ul>
    <li>alerta-server</li>
    <li>alerta-server.dist</li>
    <li>frigate</li>
    <li>gns3-gui</li>
    <li>gns3-gui.dist</li>
    <li>gns3-server</li>
    <li>gns3-server.dist</li>
    <li>kdePackages.drkonqi</li>
    <li>kdePackages.drkonqi.debug</li>
    <li>kdePackages.drkonqi.dev</li>
    <li>kdePackages.drkonqi.devtools</li>
    <li>ledfx</li>
    <li>ledfx.dist</li>
    <li>litellm</li>
    <li>litellm.dist</li>
    <li>matrix-synapse</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth.dist</li>
    <li>matrix-synapse-unwrapped</li>
    <li>matrix-synapse-unwrapped.dist</li>
    <li>moodle-dl</li>
    <li>moodle-dl.dist</li>
    <li>mozphab</li>
    <li>mozphab.dist</li>
    <li>netbox (netbox_4_2)</li>
    <li>netbox_4_1</li>
    <li>nixpkgs-manual</li>
    <li>notus-scanner (python312Packages.notus-scanner)</li>
    <li>notus-scanner.dist (python312Packages.notus-scanner.dist)</li>
    <li>octoprint</li>
    <li>octoprint.dist</li>
    <li>ospd-openvas</li>
    <li>ospd-openvas.dist</li>
    <li>peering-manager</li>
    <li>protonvpn-gui</li>
    <li>protonvpn-gui.dist</li>
    <li>python312Packages.django-rq</li>
    <li>python312Packages.django-rq.dist</li>
    <li>python312Packages.netbox-bgp</li>
    <li>python312Packages.netbox-bgp.dist</li>
    <li>python312Packages.netbox-contract</li>
    <li>python312Packages.netbox-contract.dist</li>
    <li>python312Packages.netbox-documents</li>
    <li>python312Packages.netbox-documents.dist</li>
    <li>python312Packages.netbox-floorplan-plugin</li>
    <li>python312Packages.netbox-floorplan-plugin.dist</li>
    <li>python312Packages.netbox-interface-synchronization</li>
    <li>python312Packages.netbox-interface-synchronization.dist</li>
    <li>python312Packages.netbox-napalm-plugin</li>
    <li>python312Packages.netbox-napalm-plugin.dist</li>
    <li>python312Packages.netbox-plugin-prometheus-sd</li>
    <li>python312Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python312Packages.netbox-qrcode</li>
    <li>python312Packages.netbox-qrcode.dist</li>
    <li>python312Packages.netbox-reorder-rack</li>
    <li>python312Packages.netbox-reorder-rack.dist</li>
    <li>python312Packages.netbox-routing</li>
    <li>python312Packages.netbox-routing.dist</li>
    <li>python312Packages.netbox-topology-views</li>
    <li>python312Packages.netbox-topology-views.dist</li>
    <li>python312Packages.proton-vpn-api-core</li>
    <li>python312Packages.proton-vpn-api-core.dist</li>
    <li>python312Packages.proton-vpn-network-manager</li>
    <li>python312Packages.proton-vpn-network-manager.dist</li>
    <li>python312Packages.rq</li>
    <li>python312Packages.rq.dist</li>
    <li>python312Packages.sentry-sdk (python312Packages.sentry-sdk_1)</li>
    <li>python312Packages.sentry-sdk.dist (python312Packages.sentry-sdk_1.dist)</li>
    <li>python313Packages.netbox-bgp</li>
    <li>python313Packages.netbox-bgp.dist</li>
    <li>python313Packages.netbox-documents</li>
    <li>python313Packages.netbox-documents.dist</li>
    <li>python313Packages.netbox-interface-synchronization</li>
    <li>python313Packages.netbox-interface-synchronization.dist</li>
    <li>python313Packages.netbox-plugin-prometheus-sd</li>
    <li>python313Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python313Packages.netbox-qrcode</li>
    <li>python313Packages.netbox-qrcode.dist</li>
    <li>python313Packages.netbox-reorder-rack</li>
    <li>python313Packages.netbox-reorder-rack.dist</li>
    <li>python313Packages.netbox-routing</li>
    <li>python313Packages.netbox-routing.dist</li>
    <li>recon-ng</li>
    <li>recon-ng.dist</li>
    <li>skytemple</li>
    <li>skytemple.dist</li>
    <li>tribler</li>
  </ul>
</details>

MattSturgeon commented:
> 2 packages failed to build:

wagtail-localize also fails on master, so it seems unrelated to this PR.

MattSturgeon commented:
Rebased again. I found no new instances of `Platform [!=]=`.

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 380005`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>cura</li>
    <li>netbox_3_7</li>
  </ul>
</details>
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>python312Packages.wagtail-localize</li>
    <li>python312Packages.wagtail-localize.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 95 packages built:</summary>
  <ul>
    <li>alerta-server</li>
    <li>alerta-server.dist</li>
    <li>frigate</li>
    <li>gns3-gui</li>
    <li>gns3-gui.dist</li>
    <li>gns3-server</li>
    <li>gns3-server.dist</li>
    <li>kdePackages.drkonqi</li>
    <li>kdePackages.drkonqi.debug</li>
    <li>kdePackages.drkonqi.dev</li>
    <li>kdePackages.drkonqi.devtools</li>
    <li>ledfx</li>
    <li>ledfx.dist</li>
    <li>litellm</li>
    <li>litellm.dist</li>
    <li>matrix-synapse</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth.dist</li>
    <li>matrix-synapse-unwrapped</li>
    <li>matrix-synapse-unwrapped.dist</li>
    <li>moodle-dl</li>
    <li>moodle-dl.dist</li>
    <li>mozphab</li>
    <li>mozphab.dist</li>
    <li>netbox (netbox_4_2)</li>
    <li>netbox_4_1</li>
    <li>nixpkgs-manual</li>
    <li>notus-scanner (python312Packages.notus-scanner)</li>
    <li>notus-scanner.dist (python312Packages.notus-scanner.dist)</li>
    <li>octoprint</li>
    <li>octoprint.dist</li>
    <li>ospd-openvas</li>
    <li>ospd-openvas.dist</li>
    <li>peering-manager</li>
    <li>protonvpn-gui</li>
    <li>protonvpn-gui.dist</li>
    <li>python312Packages.django-rq</li>
    <li>python312Packages.django-rq.dist</li>
    <li>python312Packages.netbox-attachments</li>
    <li>python312Packages.netbox-attachments.dist</li>
    <li>python312Packages.netbox-bgp</li>
    <li>python312Packages.netbox-bgp.dist</li>
    <li>python312Packages.netbox-contract</li>
    <li>python312Packages.netbox-contract.dist</li>
    <li>python312Packages.netbox-documents</li>
    <li>python312Packages.netbox-documents.dist</li>
    <li>python312Packages.netbox-floorplan-plugin</li>
    <li>python312Packages.netbox-floorplan-plugin.dist</li>
    <li>python312Packages.netbox-interface-synchronization</li>
    <li>python312Packages.netbox-interface-synchronization.dist</li>
    <li>python312Packages.netbox-napalm-plugin</li>
    <li>python312Packages.netbox-napalm-plugin.dist</li>
    <li>python312Packages.netbox-plugin-prometheus-sd</li>
    <li>python312Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python312Packages.netbox-qrcode</li>
    <li>python312Packages.netbox-qrcode.dist</li>
    <li>python312Packages.netbox-reorder-rack</li>
    <li>python312Packages.netbox-reorder-rack.dist</li>
    <li>python312Packages.netbox-routing</li>
    <li>python312Packages.netbox-routing.dist</li>
    <li>python312Packages.netbox-topology-views</li>
    <li>python312Packages.netbox-topology-views.dist</li>
    <li>python312Packages.proton-vpn-api-core</li>
    <li>python312Packages.proton-vpn-api-core.dist</li>
    <li>python312Packages.proton-vpn-network-manager</li>
    <li>python312Packages.proton-vpn-network-manager.dist</li>
    <li>python312Packages.rq</li>
    <li>python312Packages.rq.dist</li>
    <li>python312Packages.sentry-sdk (python312Packages.sentry-sdk_1)</li>
    <li>python312Packages.sentry-sdk.dist (python312Packages.sentry-sdk_1.dist)</li>
    <li>python313Packages.netbox-bgp</li>
    <li>python313Packages.netbox-bgp.dist</li>
    <li>python313Packages.netbox-documents</li>
    <li>python313Packages.netbox-documents.dist</li>
    <li>python313Packages.netbox-interface-synchronization</li>
    <li>python313Packages.netbox-interface-synchronization.dist</li>
    <li>python313Packages.netbox-plugin-prometheus-sd</li>
    <li>python313Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python313Packages.netbox-qrcode</li>
    <li>python313Packages.netbox-qrcode.dist</li>
    <li>python313Packages.netbox-reorder-rack</li>
    <li>python313Packages.netbox-reorder-rack.dist</li>
    <li>python313Packages.netbox-routing</li>
    <li>python313Packages.netbox-routing.dist</li>
    <li>recon-ng</li>
    <li>recon-ng.dist</li>
    <li>skytemple</li>
    <li>skytemple.dist</li>
    <li>tribler</li>
  </ul>
</details>

MattSturgeon commented:
Rebased to resolve conflicts and address new instances of `!=` and `==`.

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 380005`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>cura</li>
    <li>netbox_3_7</li>
  </ul>
</details>
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 96 packages failed to build:</summary>
  <ul>
    <li>alerta-server</li>
    <li>alerta-server.dist</li>
    <li>frigate</li>
    <li>gns3-gui</li>
    <li>gns3-gui.dist</li>
    <li>gns3-server</li>
    <li>gns3-server.dist</li>
    <li>kdePackages.drkonqi</li>
    <li>kdePackages.drkonqi.debug</li>
    <li>kdePackages.drkonqi.dev</li>
    <li>kdePackages.drkonqi.devtools</li>
    <li>ledfx</li>
    <li>ledfx.dist</li>
    <li>litellm</li>
    <li>litellm.dist</li>
    <li>matrix-synapse</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth.dist</li>
    <li>matrix-synapse-unwrapped</li>
    <li>matrix-synapse-unwrapped.dist</li>
    <li>moodle-dl</li>
    <li>moodle-dl.dist</li>
    <li>mozphab</li>
    <li>mozphab.dist</li>
    <li>netbox (netbox_4_2)</li>
    <li>netbox_4_1</li>
    <li>notus-scanner (python312Packages.notus-scanner)</li>
    <li>notus-scanner.dist (python312Packages.notus-scanner.dist)</li>
    <li>octoprint</li>
    <li>octoprint.dist</li>
    <li>ospd-openvas</li>
    <li>ospd-openvas.dist</li>
    <li>peering-manager</li>
    <li>protonvpn-gui</li>
    <li>protonvpn-gui.dist</li>
    <li>python312Packages.django-rq</li>
    <li>python312Packages.django-rq.dist</li>
    <li>python312Packages.netbox-attachments</li>
    <li>python312Packages.netbox-attachments.dist</li>
    <li>python312Packages.netbox-bgp</li>
    <li>python312Packages.netbox-bgp.dist</li>
    <li>python312Packages.netbox-contract</li>
    <li>python312Packages.netbox-contract.dist</li>
    <li>python312Packages.netbox-documents</li>
    <li>python312Packages.netbox-documents.dist</li>
    <li>python312Packages.netbox-floorplan-plugin</li>
    <li>python312Packages.netbox-floorplan-plugin.dist</li>
    <li>python312Packages.netbox-interface-synchronization</li>
    <li>python312Packages.netbox-interface-synchronization.dist</li>
    <li>python312Packages.netbox-napalm-plugin</li>
    <li>python312Packages.netbox-napalm-plugin.dist</li>
    <li>python312Packages.netbox-plugin-prometheus-sd</li>
    <li>python312Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python312Packages.netbox-qrcode</li>
    <li>python312Packages.netbox-qrcode.dist</li>
    <li>python312Packages.netbox-reorder-rack</li>
    <li>python312Packages.netbox-reorder-rack.dist</li>
    <li>python312Packages.netbox-routing</li>
    <li>python312Packages.netbox-routing.dist</li>
    <li>python312Packages.netbox-topology-views</li>
    <li>python312Packages.netbox-topology-views.dist</li>
    <li>python312Packages.proton-vpn-api-core</li>
    <li>python312Packages.proton-vpn-api-core.dist</li>
    <li>python312Packages.proton-vpn-network-manager</li>
    <li>python312Packages.proton-vpn-network-manager.dist</li>
    <li>python312Packages.rq</li>
    <li>python312Packages.rq.dist</li>
    <li>python312Packages.sentry-sdk (python312Packages.sentry-sdk_1)</li>
    <li>python312Packages.sentry-sdk.dist (python312Packages.sentry-sdk_1.dist)</li>
    <li>python312Packages.wagtail-localize</li>
    <li>python312Packages.wagtail-localize.dist</li>
    <li>python313Packages.netbox-bgp</li>
    <li>python313Packages.netbox-bgp.dist</li>
    <li>python313Packages.netbox-documents</li>
    <li>python313Packages.netbox-documents.dist</li>
    <li>python313Packages.netbox-interface-synchronization</li>
    <li>python313Packages.netbox-interface-synchronization.dist</li>
    <li>python313Packages.netbox-plugin-prometheus-sd</li>
    <li>python313Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python313Packages.netbox-qrcode</li>
    <li>python313Packages.netbox-qrcode.dist</li>
    <li>python313Packages.netbox-reorder-rack</li>
    <li>python313Packages.netbox-reorder-rack.dist</li>
    <li>python313Packages.netbox-routing</li>
    <li>python313Packages.netbox-routing.dist</li>
    <li>recon-ng</li>
    <li>recon-ng.dist</li>
    <li>skytemple</li>
    <li>skytemple.dist</li>
    <li>tribler</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

MattSturgeon commented:
Rebased again due to conflict with #398403

Not sure why there's so many build failures since the last rebase. The ones I've checked on master also fail there, but I've only tried a handful on master so far.

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 380005`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages marked as broken and skipped:</summary>
  <ul>
    <li>cura</li>
    <li>netbox_3_7</li>
  </ul>
</details>
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 53 packages failed to build:</summary>
  <ul>
    <li>litellm</li>
    <li>litellm.dist</li>
    <li>netbox (netbox_4_2)</li>
    <li>netbox_4_1</li>
    <li>octoprint</li>
    <li>octoprint.dist</li>
    <li>peering-manager</li>
    <li>python312Packages.django-rq</li>
    <li>python312Packages.django-rq.dist</li>
    <li>python312Packages.netbox-attachments</li>
    <li>python312Packages.netbox-attachments.dist</li>
    <li>python312Packages.netbox-bgp</li>
    <li>python312Packages.netbox-bgp.dist</li>
    <li>python312Packages.netbox-contract</li>
    <li>python312Packages.netbox-contract.dist</li>
    <li>python312Packages.netbox-documents</li>
    <li>python312Packages.netbox-documents.dist</li>
    <li>python312Packages.netbox-floorplan-plugin</li>
    <li>python312Packages.netbox-floorplan-plugin.dist</li>
    <li>python312Packages.netbox-interface-synchronization</li>
    <li>python312Packages.netbox-interface-synchronization.dist</li>
    <li>python312Packages.netbox-napalm-plugin</li>
    <li>python312Packages.netbox-napalm-plugin.dist</li>
    <li>python312Packages.netbox-plugin-prometheus-sd</li>
    <li>python312Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python312Packages.netbox-qrcode</li>
    <li>python312Packages.netbox-qrcode.dist</li>
    <li>python312Packages.netbox-reorder-rack</li>
    <li>python312Packages.netbox-reorder-rack.dist</li>
    <li>python312Packages.netbox-routing</li>
    <li>python312Packages.netbox-routing.dist</li>
    <li>python312Packages.netbox-topology-views</li>
    <li>python312Packages.netbox-topology-views.dist</li>
    <li>python312Packages.rq</li>
    <li>python312Packages.rq.dist</li>
    <li>python312Packages.wagtail-localize</li>
    <li>python312Packages.wagtail-localize.dist</li>
    <li>python313Packages.netbox-bgp</li>
    <li>python313Packages.netbox-bgp.dist</li>
    <li>python313Packages.netbox-documents</li>
    <li>python313Packages.netbox-documents.dist</li>
    <li>python313Packages.netbox-interface-synchronization</li>
    <li>python313Packages.netbox-interface-synchronization.dist</li>
    <li>python313Packages.netbox-plugin-prometheus-sd</li>
    <li>python313Packages.netbox-plugin-prometheus-sd.dist</li>
    <li>python313Packages.netbox-qrcode</li>
    <li>python313Packages.netbox-qrcode.dist</li>
    <li>python313Packages.netbox-reorder-rack</li>
    <li>python313Packages.netbox-reorder-rack.dist</li>
    <li>python313Packages.netbox-routing</li>
    <li>python313Packages.netbox-routing.dist</li>
    <li>recon-ng</li>
    <li>recon-ng.dist</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 44 packages built:</summary>
  <ul>
    <li>alerta-server</li>
    <li>alerta-server.dist</li>
    <li>frigate</li>
    <li>gns3-gui</li>
    <li>gns3-gui.dist</li>
    <li>gns3-server</li>
    <li>gns3-server.dist</li>
    <li>kdePackages.drkonqi</li>
    <li>kdePackages.drkonqi.debug</li>
    <li>kdePackages.drkonqi.dev</li>
    <li>kdePackages.drkonqi.devtools</li>
    <li>ledfx</li>
    <li>ledfx.dist</li>
    <li>matrix-synapse</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider</li>
    <li>matrix-synapse-plugins.matrix-synapse-s3-storage-provider.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth.dist</li>
    <li>matrix-synapse-unwrapped</li>
    <li>matrix-synapse-unwrapped.dist</li>
    <li>moodle-dl</li>
    <li>moodle-dl.dist</li>
    <li>mozphab</li>
    <li>mozphab.dist</li>
    <li>nixpkgs-manual</li>
    <li>notus-scanner (python312Packages.notus-scanner)</li>
    <li>notus-scanner.dist (python312Packages.notus-scanner.dist)</li>
    <li>ospd-openvas</li>
    <li>ospd-openvas.dist</li>
    <li>protonvpn-gui</li>
    <li>protonvpn-gui.dist</li>
    <li>python312Packages.proton-vpn-api-core</li>
    <li>python312Packages.proton-vpn-api-core.dist</li>
    <li>python312Packages.proton-vpn-network-manager</li>
    <li>python312Packages.proton-vpn-network-manager.dist</li>
    <li>python312Packages.sentry-sdk (python312Packages.sentry-sdk_1)</li>
    <li>python312Packages.sentry-sdk.dist (python312Packages.sentry-sdk_1.dist)</li>
    <li>skytemple</li>
    <li>skytemple.dist</li>
    <li>tribler</li>
  </ul>
</details>

FliegendeWurst reviewed (approved):
Another merge conflict appeared. LGTM otherwise.

emilazy commented:
Sorry for taking so long to get around to this.

My tl;dr take:

* It’s unfortunate that Nix `==` behaves so incoherently for equality on objects‐with‐methods like this. The hack to cause a warning or error is cool, but short‐circuiting means it will never catch things reliably. That makes me inclined to approaches that separate out data and functions where possible, as sad a limitation as that is.

* I also think it’s valuable to be able to serialize platform definitions to JSON, etc.

* However, you’re the ones who have put a ton of work into cleaning this stuff up, and I don’t know if there’s any easy way to get the desired behaviour with my preferred approach, so I don’t want my cheap takes to override your hard work; this is just my 2¢.

* In the long run, I quite strongly believe that we should separate out *inputs* to platform elaboration from *outputs*. Currently we have one monolithic structure that gets “mutated” with outputs, and some things you specify on input get ignored, some produce incoherent combinations, and so on. It seems to me that we need separate types and interfaces for *things you can use to specify a platform to Nixpkgs* and *computed values Nixpkgs derives from a platform specification*. There will be overlap, of course, but to me this lack of separation is the root cause of all the problems we’ve had around this. I believe addressing this would probably allow us to have a plain‐old‐data approach while being able to adjust platform configurations by modifying the inputs rather than the outputs.

* Honestly I’m not sure how much actually comparing platforms for exact equality makes sense as an operation in general…

FliegendeWurst commented on doc/release-notes/rl-2505.section.md:
```diff
@@ -41,6 +41,10 @@
 
+- Using `==` or `!=` to compare elaborated platform attrsets, such as `stdenv.hostPlatform` and `stdenv.buildPlatform`, is no longer supported and may produce a warning. In a future version, this warning will be replaced with an error. Instead, you should use the `canExecute`, `equals`, or `notEquals` functions included with platform attrsets or use `lib.systems.equals`.
+
+  - E.g. replace `hostPlatform == buildPlatform` with `buildPlatform.canExecute hostPlatform`.
```
I know of many packages where the build system just doesn't understand the distinction between build and host. For such cases an equality check makes sense.

MattSturgeon commented:
> Sorry for taking so long to get around to this.

> I don’t want my cheap takes to override your hard work; this is just my 2¢.

I for one value your input! :heart: 

---

> * It’s unfortunate that Nix `==` behaves so incoherently for equality on objects‐with‐methods like this.

Agreed. It's also worth noting that `==` is not _just_ an issue because of functions, but also because of wanting to do things like separate input and output attrs within the platform. Two platforms with the same outputs should be considered equal regardless of their inputs.

---

> The hack to cause a warning or error is cool, but short‐circuiting means it will never catch things reliably.

Yes, it's sad that we can't find a reliable way to print a warning on `==` use, but I think it is good enough to catch most cases.
 
> * I also think it’s valuable to be able to serialize platform definitions to JSON, etc.

This would be possible with a `lib.systems.toJSON` function that uses the same `removeUnwanted` as `equals` does. Sad that you can't _directly_ pass the platform to `builtins.toJSON`, but I don't see a way around this.

----

> That makes me inclined to approaches that separate out data and functions where possible, as sad a limitation as that is.

> * In the long run, I quite strongly believe that we should separate out _inputs_ to platform elaboration from _outputs_.

That would _also_ require the same `==` deprecation, assuming inputs and outputs were both stored within the `platform` structure.

#380342 is attempting to do something _similar_ by capturing inputs in the `override` function's scope.

> * Honestly I’m not sure how much actually comparing platforms for exact equality makes sense as an operation in general…

:+1:

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 380005`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>frigate</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>nixpkgs-manual</li>
  </ul>
</details>

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 380005`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>frigate</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

emilazy commented:
> That would _also_ require the same `==` deprecation, assuming inputs and outputs were both stored within the `platform` structure.

(FWIW, explicitly *not* mixing them was my intention – it makes sense to me to separate inputs and computed values, because that is exactly what lets you compute new values from a modified input. However I realize that would be a considerably bigger change to disentangle all this stuff.)
